server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1115,7 +1115,18 @@ public void snapshotShard(Store store, MapperService mapperService, SnapshotId s
                     protected void doRun() {
                         try {
                             if (alreadyFailed.get() == false) {
-                                snapshotFile(snapshotFileInfo, indexId, shardId, snapshotId, snapshotStatus, store);
+                                if (store.tryIncRef()) {
```

Comment:
This whole loop is kinda awkward to begin with ... makes me wonder if we shouldn't just run this on the generic pool and makethe parallelism for snapshots configurable explicitly exactly like we do for recoveries ...

Reply:
> makethe parallelism for snapshots configurable explicitly exactly like we do for recoveries

Not sure to follow you :(

Reply:
Sorry badly explained :)

I find this whole loop over all the files really strange. We currently create one `Runnable` for each file to upload individually then enqueue all the runnables. That forces us to do the strange `alreadyFailed` flag to not get crazy exceptions and also to increment and decrement the ref count on the store for each file individually.
It seems like it would be more correct/simpler and less hacky to simply have a queue of files and have worker pull from that queue until its empty. Then each worker can just get that reference once and we don't have to run all `N` tasks for `N` files even if the first file fails uploading.

Reply:
Thanks for explaining, it makes sense but I don't see this as a requirement to merge this PR. Let's keep this in our mind for the rainy boring days ;)

Reply:
Yea, this was more of a general comment to justify the weird code :)